src/plugins/lti/LTI.jsx

```diff
@@ -0,0 +1,140 @@
+import React, { useState, useEffect, useContext } from "react"
+import shortid from "shortid"
+
+import UserContext from "~/Contexts/User"
+
+import Flex from "~/components/Flex"
+import Loader from "~/components/Loader"
+import { getTools, getPseudonym, sign, depseudonymizationFrame } from './utils'
+
+const LTI = ({ state, editable }) => {
+  const [tools, setTools] = useState({})
+  const [loading, setLoading] = useState(true)
+  const [idToken, setIdToken] = useState("")
+  const { store: user } = useContext(UserContext)
+
+  async function bootstrap() {
+    if (state.url.value) {
+      const iss = window.location.protocol + '//' + window.location.host + '/';
+      const pseudonym = await getPseudonym(user.id, state.templateId.value)
+      const current = new Date();
+      const name = encodeURI(depseudonymizationFrame(iss, pseudonym));
+      if(state.ltiVersion.value === '1.3.0') {
+        const request = {
+          iss,
+          name,
+          aud: state.clientId.value,
+          sub: pseudonym,
+          exp: current.getTime() + 3 * 60,
+          iat: current.getTime(),
+          nonce: shortid.generate() + shortid.generate(),
+          'https://purl.imsglobal.org/spec/lti/claim/message_type': state.ltiMessageType.value,
+          'https://purl.imsglobal.org/spec/lti/claim/roles': [
+            `http://purl.imsglobal.org/vocab/lis/v2/membership#${user.roles[0].name}`,
+          ],
+          'https://purl.imsglobal.org/spec/lti/claim/resource_link': {
+            id: state.id.value,
+          },
+          'https://purl.imsglobal.org/spec/lti/claim/version': state.ltiVersion.value,
+          'https://purl.imsglobal.org/spec/lti/claim/deployment_id': state.id.value,
+          'https://purl.imsglobal.org/spec/lti-dl/claim/deep_linking_settings':
+            (state.ltiMessageType.value === 'LtiDeepLinkingRequest'
+              ? {
+                accept_types: ['ltiLink'],
+                accept_media_types: 'image/*,text/html',
+                accept_presentation_document_targets: ['iframe', 'window'],
+                deep_link_return_url: `http://localhost:3030/tools/${state.id.value}/link`,
+              }
+              : undefined),
+        }
+        setIdToken(await sign(request))
+        window["link-" + state.id.value] = (url) => {
+          state.ltiMessageType.set('basic-lti-launch-request')
+          state.url.set(url)
+          setIdToken(null)
+        }
+      }
+    } else {
+      setTools(await getTools())
+    }
+    setLoading(false)
+  }
+
+  useEffect(() => {
+      bootstrap()
+  }, [])
+
+  const setTool = (tool) => {
+    state.id.set(shortid.generate())
+    state.templateId.set(tool._id)
+    state.clientId.set(tool.oAuthClientId)
+    state.ltiMessageType.set(tool.lti_message_type)
+    state.ltiVersion.set(tool.lti_version)
+    state.url.set(tool.url)
+  }
+
+  if (loading) {
```

Comment:
Only style, but maybe the if / else render is more readable if you use named helper for the render block.
const renderDefault = () => return ( <div> ..
if() renderDefault();  ...

Reply:
Thanks!